File: extensions/cppClientAPI/eccCommonDataModel/Metrics/Entity.cpp
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2013
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          21-June-2013
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

#include "stdafx.h"
#include "Entity.h"

using namespace boost::uuids;
using namespace std;




namespace ecc_commonDataModel
{

Entity::Entity()
{
  entityUniqueID = createRandomUUID();
}
    
Entity::Entity( Entity::ptr_t ent )
{
  if ( ent )
  {
    entityUniqueID    = ent->getUUID();
    entityID          = ent->getEntityID();
    entityName        = ent->getName();
    entityDescription = ent->getDescription();

    Attribute::Set srcAtts = ent->getAttributes();
    entityAttributes.insert( srcAtts.begin(), srcAtts.end() );
  }
}
    
Entity::Entity( const UUID& uuid )
{
  entityUniqueID = uuid;
}

Entity::Entity( const UUID&   uuid, 
                const String& name, 
                const String& description )
{
  entityUniqueID    = uuid;
  entityName        = name;
  entityDescription = description;
}
    
Entity::Entity( const UUID&           uuid, 
                const String&         name, 
                const String&         description, 
                const Attribute::Set& atts )
{
  entityUniqueID    = uuid;
  entityName        = name;
  entityDescription = description;

  Attribute::Set::const_iterator attIt = atts.begin();
  while ( attIt != atts.end() )
  {
    Attribute::ptr_t attr = *attIt;

    if ( attr ) entityAttributes.insert( attr );

    ++attIt;
  }
}
    
Entity::Entity( const UUID&   uuid, 
                const String& entID, 
                const String& name, 
                const String& description )
{
  entityUniqueID    = uuid;
  entityID          = entID;
  entityName        = name;
  entityDescription = description;
}
    
Entity::Entity( const UUID&           uuid, 
                const String&         entID, 
                const String&         name, 
                const String&         description, 
                const Attribute::Set& atts )
{
  entityUniqueID    = uuid;
  entityID          = entID;
  entityName        = name;
  entityDescription = description;

  Attribute::Set::const_iterator attIt = atts.begin();
  while ( attIt != atts.end() )
  {
    Attribute::ptr_t attr = *attIt;

    if ( attr ) entityAttributes.insert( attr );

    ++attIt;
  }
}

Entity::~Entity()
{
}

UUID Entity::getUUID()
{
  return entityUniqueID;
}

void Entity::setUUID( const UUID& ID )
{
  entityUniqueID = ID;
}

String Entity::getEntityID()
{
  return entityID;
}

void Entity::setEntityID( const String& ID )
{
  entityID = ID;
}

String Entity::getName()
{
  return entityName;
}

void Entity::setName( const String& name )
{
  entityName = name;
}

String Entity::getDescription()
{
  return entityDescription;
}

void Entity::setDescription( const String& desc )
{
  entityDescription = desc;
}

Attribute::Set Entity::getAttributes()
{
  return entityAttributes;
}
  
void Entity::setAttributes( const Attribute::Set& attributes )
{
  entityAttributes = attributes;
}
    
void Entity::addAttribute( Attribute::ptr_t attribute )
{
  if ( attribute )
    entityAttributes.insert( attribute );
}
    
void Entity::addAttributes( const Attribute::Set& attributes )
{
  Attribute::Set::const_iterator attIt = attributes.begin();

  while ( attIt != attributes.end() )
  {
    Attribute::ptr_t attr = *attIt;

    if ( attr ) entityAttributes.insert( attr );

    ++attIt;
  }
}

// ModelBase -----------------------------------------------------------------
String Entity::toJSON()
{
  String json( L"{" );

  json.append( createJSON_Prop( L"uuid", uuidToWide(entityUniqueID) ) + L"," );

  json.append( createJSON_Prop( L"entityID", entityID ) + L"," );

  json.append( createJSON_Prop( L"name", entityName ) + L"," );

  json.append( createJSON_Prop( L"description", entityDescription ) + L"," );

  // Attributes
  json.append( L"\"attributes\":[" );

  Attribute::Set::const_iterator atIt = entityAttributes.begin();
  while ( atIt != entityAttributes.end() )
  {
    json.append( (*atIt)->toJSON() + L"," );

    ++atIt;
  }

  // Snip off trailing delimiter
  if ( !entityAttributes.empty() )
  {
    unsigned int jLen = json.length();
    json = json.substr( 0, jLen-1 );
  }

  json.append( L"]" );

  json.append( L"}" );

  return json;
}

void Entity::fromJSON( const ModelBase::JSONTree& jsonTree )
{
  // Client does not require implementation
}

String Entity::toString()
{
  return entityName + L" {" + uuidToWide(entityUniqueID) + L"}";
}

} // namespace

File: common/dataModel/metrics/src/main/java/uk/ac/soton/itinnovation/experimedia/arch/ecc/common/dataModel/metrics/Entity.java
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2012
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Vegard Engen
//      Created Date :          2012-08-09
//      Created for Project :   BonFIRE
//
/////////////////////////////////////////////////////////////////////////
package uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.metrics;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

/**
 * An experiment can consist of many different entities that are monitored, some
 * of which could be system components, virtual resources or human beings. Each of
 * which have attributes that can be monitored, which are specified for each entity.
 * 
 * An entity can be described in terms of a name, description and a unique identifier.
 * It is also described in terms of observable attributes that can be monitored.
 * Any measurements produced by metric generators will refer to a specific attribute
 * of an entity.
 * 
 * Examples of entities include:
 *   - A person
 *   - A system component / service that could be used in many experiments
 *   - A virtual compute resource
 * 
 * @author Vegard Engen
 */
public class Entity implements Serializable
{
    private UUID uuid; // used to uniquely identify an entity in our framework
    private String entityID; // can be used to identify the entity according to 'external' ID
    private String name;
    private String description;
    private Set<Attribute> attributes;
    
    // other attributes needed?
    //   - identifier such as URI? 
    //   - type of entity
    //   - specification of that entity etc (if VM, then instance type etc)
    //   - add a Map of key-value pairs of things for such general spec information?

    /**
     * Default constructor, which creates a random UUID for this object instance
     * and initialises the attributes and experimentUUIDs sets.
     */
    public Entity()
    {
        this.uuid = UUID.randomUUID();
        this.attributes = new HashSet<Attribute>();
    }
    
    /**
     * A copy constructor; takes a deep copy of the attributes and experiment UUIDs.
     * @param e An Entity object from which a copy is made.
     */
    public Entity(Entity e)
    {
        if (e == null)
            return;
        
        if (e.getUUID() != null)
            this.uuid = UUID.fromString(e.getUUID().toString());
        this.name = e.getName();
        this.description = e.getDescription();
        this.entityID = e.getEntityID();
        
        this.attributes = new HashSet<Attribute>();
        if (e.getAttributes() != null)
        {
            for (Attribute a : e.getAttributes())
            {
                if (a != null)
                    this.attributes.add(new Attribute(a));
            }
        }
    }
    
    /**
     * A constructor to set all the fields of the entity class except for the attributes.
     * @param uuid A UUID used to uniquely identify an entity in this framework.
     */
    public Entity(UUID uuid)
    {
        this();
        this.uuid = uuid;
    }
    
    /**
     * A constructor to set all the fields of the entity class except for the attributes.
     * @param uuid A UUID used to uniquely identify an entity in this framework.
     * @param name The name of the Entity.
     * @param description A description of the entity.
     */
    public Entity(UUID uuid, String name, String description)
    {
        this(uuid);
        this.name = name;
        this.description = description;
    }
    
    /**
     * A constructor to set all the fields of the entity class.
     * @param uuid A UUID used to uniquely identify an entity in this framework.
     * @param name The name of the Entity.
     * @param description A description of the entity.
     * @param attributes A set of attributes of the entity, which could be observed to generate metrics.
     */
    public Entity(UUID uuid, String name, String description, Set<Attribute> attributes)
    {
        this(uuid, name, description);
        this.attributes = attributes;
    }
    
    /**
     * A constructor to set all the fields of the entity class except for the attributes.
     * @param uuid A UUID used to uniquely identify an entity in this framework.
     * @param entityID An ID that can be used if the Entity is known by a particular ID outside of the framework.
     * @param name The name of the Entity.
     * @param description A description of the entity.
     */
    public Entity(UUID uuid, String entityID, String name, String description)
    {
        this(uuid);
        this.name = name;
        this.description = description;
        this.entityID = entityID;
    }
    
    /**
     * A constructor to set all the fields of the entity class.
     * @param uuid A UUID used to uniquely identify an entity in this framework.
     * @param entityID An ID that can be used if the Entity is known by a particular ID outside of the framework.
     * @param name The name of the Entity.
     * @param description A description of the entity.
     * @param attributes A set of attributes of the entity, which could be observed to generate metrics.
     */
    public Entity(UUID uuid, String entityID, String name, String description, Set<Attribute> attributes)
    {
        this(uuid, entityID, name, description);
        this.attributes = attributes;
    }
    
    /**
     * @return the uuid
     */
    public UUID getUUID()
    {
        return uuid;
    }

    /**
     * @param uuid the uuid to set
     */
    public void setUUID(UUID uuid)
    {
        this.uuid = uuid;
    }

    /**
     * @return the name
     */
    public String getName()
    {
        return name;
    }

    /**
     * @param name the name to set
     */
    public void setName(String name)
    {
        this.name = name;
    }

    /**
     * @return the description
     */
    public String getDescription()
    {
        return description;
    }

    /**
     * @param description the description to set
     */
    public void setDescription(String description)
    {
        this.description = description;
    }

    /**
     * @return the attributes
     */
    public Set<Attribute> getAttributes()
    {
        return attributes;
    }

    /**
     * @param attributes the attributes to set
     */
    public void setAttributes(Set<Attribute> attributes)
    {
        this.attributes = attributes;
    }
    
    /**
     * @param attribute the attribute to add
     */
    public void addAttribute(Attribute attribute)
    {
        if (attribute == null)
            return;
        
        if (this.attributes == null)
            this.attributes = new HashSet<Attribute>();
        
        this.attributes.add(attribute);
    }
    
    /**
     * @param attributes the attributes to add
     */
    public void addAttributes(Set<Attribute> attributes)
    {
        if (attributes == null || attributes.isEmpty())
            return;
        
        if (this.attributes == null)
            this.attributes = new HashSet<Attribute>();
        
        this.attributes.addAll(attributes);
    }
    
    @Override
    public String toString()
    {
        return name;
    }

    /**
     * @return the entityID
     */
    public String getEntityID()
    {
        return entityID;
    }

    /**
     * @param entityID the entityID to set
     */
    public void setEntityID(String entityID)
    {
        this.entityID = entityID;
    }

    /**
     * @return the experimentUUIDs
     *
    public Set<UUID> getExperimentUUIDs()
    {
        return experimentUUIDs;
    }*/

    /**
     * @param experimentUUIDs the experimentUUIDs to set
     *
    public void setExperimentUUIDs(Set<UUID> experimentUUIDs)
    {
        this.experimentUUIDs = experimentUUIDs;
    }*/
    
    /**
     * @param experimentUUID the experimentUUID to add
     *
    public void addExperimentUUID(UUID experimentUUID)
    {
        if (experimentUUID == null)
            return;
        
        if (this.experimentUUIDs == null)
            this.experimentUUIDs = new HashSet<UUID>();
        
        this.experimentUUIDs.add(experimentUUID);
    }*/
    
    /**
     * @param experimentUUIDs the experimentUUIDs to add
     *
    public void addExperimentUUIDs(Set<UUID> experimentUUIDs)
    {
        if ((experimentUUIDs == null) || experimentUUIDs.isEmpty())
            return;
        
        if (this.experimentUUIDs == null)
            this.experimentUUIDs = new HashSet<UUID>();
        
        this.experimentUUIDs.addAll(experimentUUIDs);
    }*/
}
